drivers/adc/adc_mcux_adc16.c

```diff
@@ -305,3 +305,13 @@ static const struct adc_driver_api mcux_adc16_driver_api = {
 	}
 
 DT_INST_FOREACH(ACD16_MCUX_INIT)
+
+#define SUBSYS_DEVICE_REGISTER(subsys, name) \
```

Comment:
I would see such registration being related to the shell, so it would not register devices (and thus take ROM) if nothing is using this. (if we get other use case, we'll enable the registration for these as well later on)
if device shell module is enabled: all devices would be registered.
If only i2c shell module is enabled, only i2c devices would be registered. (same for sensors etc...)
i2c registration would check whether device or i2c shell is enabled.

That being said:
- I would not follow that path _unless_ we get rid of the device names. Let's not add a feature replacing (and improving) an existing one, if we cannot remove the old one.
- If we do such registration, let's make it simple: DEVICE_DEFINE() would take the subsys parameter. (Note that I am talking about the right DEVICE_DEFINE we should have someday: that will be unique, that will transparently handle PM, and which would not have competitors such as DEVICE_AND_API_INIT which one would not exist anymore)